File: forecast/tests/fakes.py
"""測試用的假物件與資料產生器：不連網、不連資料庫，也不依賴被 .gitignore 的 samples/。"""
from datetime import datetime, timedelta, timezone

from tw_forecast.frontend.regions import CITY_ORDER

TZ = timezone(timedelta(hours=8))


# ---------- 假 Supabase ----------
class FakeResult:
    def __init__(self, data):
        self.data = data


class FakeQuery:
    """支援 select／lte／lt／gte／gt／eq／in_／order／limit／execute 與 upsert，資料放在記憶體。"""

    def __init__(self, table):
        self._table = table
        self.rows = list(table.rows)
        self._limit = None
        self._cols = None
        self._upsert = None

    def select(self, cols="*"):
        self._cols = None if cols == "*" else cols.split(",")
        return self

    def _cmp(self, col, value, fn):
        v = datetime.fromisoformat(value)
        self.rows = [r for r in self.rows if fn(datetime.fromisoformat(r[col]), v)]
        return self

    def lte(self, col, value): return self._cmp(col, value, lambda a, b: a <= b)
    def lt(self, col, value): return self._cmp(col, value, lambda a, b: a < b)
    def gte(self, col, value): return self._cmp(col, value, lambda a, b: a >= b)
    def gt(self, col, value): return self._cmp(col, value, lambda a, b: a > b)

    def eq(self, col, value):
        self.rows = [r for r in self.rows if r[col] == value]
        return self

    def in_(self, col, values):
        self.rows = [r for r in self.rows if r[col] in values]
        return self

    def order(self, col, desc=False):
        self.rows.sort(key=lambda r: datetime.fromisoformat(r[col]), reverse=desc)
        return self

    def limit(self, n):
        self._limit = n
        return self

    def upsert(self, payload, on_conflict=None):
        self._upsert = (payload, on_conflict)
        return self

    def execute(self):
        if self._upsert is not None:
            payload, on_conflict = self._upsert
            self._table.upserts.append((payload, on_conflict))
            return FakeResult(payload)
        rows = self.rows[: self._limit] if self._limit else self.rows
        if self._cols:
            rows = [{k: r[k] for k in self._cols} for r in rows]
        return FakeResult([dict(r) for r in rows])


class FakeTable:
    def __init__(self, rows=None):
        self.rows = list(rows or [])
        self.upserts = []  # 記錄每次 upsert 的 (資料, on_conflict)，測試用來斷言


class FakeClient:
    """table(name) 回傳查詢物件；rpc(name, params) 回傳預先設定好的結果（或拋出預先設定的例外）。"""

    def __init__(self, tables=None, rpc_results=None):
        self.tables = {name: FakeTable(rows) for name, rows in (tables or {}).items()}
        self.rpc_results = rpc_results or {}
        self.rpc_calls = []

    def table(self, name):
        return FakeQuery(self.tables.setdefault(name, FakeTable()))

    def rpc(self, name, params):
        self.rpc_calls.append((name, params))
        result = self.rpc_results[name]
        if isinstance(result, Exception):
            raise result
        return _RpcCall(result)


class _RpcCall:
    def __init__(self, data):
        self._data = data

    def execute(self):
        return FakeResult(self._data)


# ---------- 假氣象署 API 回應 ----------
def cwa_payload(city="臺北市", start="2026-09-21T06:00:00+08:00", end="2026-09-21T18:00:00+08:00", **override):
    """一個縣市、一個時段的最小 API 回應。override 可替換要素的值（如 MaxTemperature="-"）。"""
    values = {"Weather": "多雲", "MinTemperature": "25", "MaxTemperature": "31", "Temperature": "28",
              "ProbabilityOfPrecipitation": "30", "MaxComfortIndexDescription": "悶熱"}
    values.update(override)

    def element(name, key):
        return {"ElementName": name, "Time": [{"StartTime": start, "EndTime": end, "ElementValue": [{key: values[key]}]}]}

    return {"success": "true", "records": {"Locations": [{"Location": [{
        "LocationName": city, "Latitude": "25.05", "Longitude": "121.5",
        "WeatherElement": [
            element("天氣現象", "Weather"), element("最低溫度", "MinTemperature"),
            element("最高溫度", "MaxTemperature"), element("平均溫度", "Temperature"),
            element("12小時降雨機率", "ProbabilityOfPrecipitation"),
            element("最大舒適度指數", "MaxComfortIndexDescription")]}]}]}}


# ---------- 假預報資料庫列（22 縣市 × 多個 12 小時時段） ----------
def forecast_rows(first_day: datetime, days: int = 4, updated_at: str = "2026-09-21T09:30:00+08:00") -> list[dict]:
    """每個縣市、每天 06:00~18:00 與 18:00~隔天 06:00 兩個完整時段。溫度隨縣市與時段變化，方便驗證極值。"""
    rows = []
    for c, city in enumerate(CITY_ORDER):
        for d in range(days * 2):
            start = first_day + timedelta(hours=12 * d)
            rows.append({
                "location_name": city,
                "forecast_time_start": start.isoformat(), "forecast_time_end": (start + timedelta(hours=12)).isoformat(),
                "latitude": 22.0 + c * 0.15, "longitude": 120.0 + c * 0.08,
                "weather_condition": ["晴", "多雲", "陰短暫雨"][d % 3],
                "min_temp": 18.0 + c * 0.5, "max_temp": 24.0 + c * 0.5 + d, "avg_temp": 21.0 + c * 0.5,
                "rain_probability": None if d > 5 else (d * 10 + c) % 100,
                "comfort_index": "舒適", "updated_at": updated_at,
            })
    return rows

File: forecast/web/tests/fakes.ts
/** 測試用的假物件與資料產生器：不連網、不連資料庫（對應 Python 的 tests/fakes.py）。 */
import type { SupabaseClient } from "@supabase/supabase-js";

import { CITY_ORDER } from "../src/lib/regions";

type Row = Record<string, unknown>;

/** 台灣時間的 Date，例如 tw(2026, 9, 21, 6) = 2026-09-21 06:00 (+08:00)。 */
export function tw(year: number, month: number, day: number, hour = 0, minute = 0): Date {
  const p = (n: number) => String(n).padStart(2, "0");
  return new Date(`${year}-${p(month)}-${p(day)}T${p(hour)}:${p(minute)}:00+08:00`);
}

const time = (v: unknown) => new Date(String(v)).getTime();

/** 模仿 supabase-js 的查詢串接（select／lte／lt／gte／gt／eq／in／order／limit），await 時回傳 { data, error }。 */
class FakeQuery implements PromiseLike<{ data: Row[] | null; error: { message: string } | null }> {
  private rows: Row[];
  private cols: string[] | null = null;
  private max: number | null = null;

  constructor(rows: Row[], private readonly error: string | null) {
    this.rows = [...rows];
  }

  select(cols = "*") {
    this.cols = cols === "*" ? null : cols.split(",");
    return this;
  }

  private cmp(col: string, value: string, fn: (a: number, b: number) => boolean) {
    this.rows = this.rows.filter((r) => fn(time(r[col]), time(value)));
    return this;
  }

  lte(col: string, v: string) { return this.cmp(col, v, (a, b) => a <= b); }
  lt(col: string, v: string) { return this.cmp(col, v, (a, b) => a < b); }
  gte(col: string, v: string) { return this.cmp(col, v, (a, b) => a >= b); }
  gt(col: string, v: string) { return this.cmp(col, v, (a, b) => a > b); }

  eq(col: string, v: unknown) {
    this.rows = this.rows.filter((r) => r[col] === v);
    return this;
  }

  in(col: string, values: unknown[]) {
    this.rows = this.rows.filter((r) => values.includes(r[col]));
    return this;
  }

  order(col: string, { ascending = true } = {}) {
    this.rows.sort((a, b) => (ascending ? 1 : -1) * (time(a[col]) - time(b[col])));
    return this;
  }

  limit(n: number) {
    this.max = n;
    return this;
  }

  then<A, B>(
    ok?: ((v: { data: Row[] | null; error: { message: string } | null }) => A | PromiseLike<A>) | null,
    fail?: ((reason: unknown) => B | PromiseLike<B>) | null,
  ): PromiseLike<A | B> {
    let rows = this.max === null ? this.rows : this.rows.slice(0, this.max);
    if (this.cols) rows = rows.map((r) => Object.fromEntries(this.cols!.map((c) => [c, r[c]])));
    const result = this.error
      ? { data: null, error: { message: this.error } }
      : { data: rows.map((r) => ({ ...r })), error: null };
    return Promise.resolve(result).then(ok, fail);
  }
}

/** 假 Supabase：tables 為各表的資料；errors 指定某張表的查詢要回傳錯誤。 */
export function fakeClient(tables: Record<string, Row[]>, errors: Record<string, string> = {}): SupabaseClient {
  return { from: (name: string) => new FakeQuery(tables[name] ?? [], errors[name] ?? null) } as unknown as SupabaseClient;
}

/** 假預報資料：每個縣市、每天 06:00~18:00 與 18:00~隔天 06:00 兩個完整時段。溫度隨縣市與時段變化。 */
export function forecastRows(firstDay: Date, days = 4, updatedAt = "2026-09-21T09:30:00+08:00"): Row[] {
  const rows: Row[] = [];
  CITY_ORDER.forEach((city, c) => {
    for (let d = 0; d < days * 2; d++) {
      const start = new Date(firstDay.getTime() + d * 12 * 3600_000);
      rows.push({
        location_name: city,
        forecast_time_start: start.toISOString(),
        forecast_time_end: new Date(start.getTime() + 12 * 3600_000).toISOString(),
        latitude: 22.0 + c * 0.15, longitude: 120.0 + c * 0.08,
        weather_condition: ["晴", "多雲", "陰短暫雨"][d % 3],
        min_temp: 18.0 + c * 0.5, max_temp: 24.0 + c * 0.5 + d, avg_temp: 21.0 + c * 0.5,
        rain_probability: d > 5 ? null : (d * 10 + c) % 100,
        comfort_index: "舒適", updated_at: updatedAt,
      });
    }
  });
  return rows;
}
